qiskit/circuit/add_control.py

```diff
@@ -92,6 +93,40 @@ def control(operation: Union[Gate, ControlledGate],
     import qiskit.extensions.standard.multi_control_toffoli_gate
     import qiskit.extensions.standard.multi_control_u1_gate
 
+    if isinstance(operation, controlledgate.ControlledGate):
+        new_num_ctrl_qubits = num_ctrl_qubits + operation.num_ctrl_qubits
+        base_name = operation.base_gate.name
+        base_gate = operation.base_gate
+    else:
+        new_num_ctrl_qubits = num_ctrl_qubits
+        base_name = operation.name
+        base_gate = operation
+    # In order to maintain some backward compatibility with gate names this
+    # uses a naming convention where if the number of controls is <=2 the gate
+    # is named like "cc<base_gate.name>", else it is named like
+    # "c<num_ctrl_qubits><base_name>".
+    if new_num_ctrl_qubits > 2:
+        ctrl_substr = 'c{0:d}'.format(new_num_ctrl_qubits)
+    else:
+        ctrl_substr = ('{0}' * new_num_ctrl_qubits).format('c')
+    new_name = '{0}{1}'.format(ctrl_substr, base_name)
+
+    # define controlled gate instructions
+
+    if (not operation.definition and
+            not isinstance(operation, (CXGate, U3Gate, IGate))):
```

Comment:
Is it possible to avoid hardcoding these as a means of deciding whether something is opaque or not? 

Since CXGate, U3Gate, etc. have some rules defined for how they should be controlled (e.g. CCX, MCT, ..), then I think it should be possible to consult those rules first, if nothing found, then try to decompose the gate according to its definition, and if that doesn't exist, then treat it as opaque.

Soon CXGate, U3Gate will not be "base" of all gates, they will themselves have rewriting rules for example to convert to CZGate.

Reply:
I was able to avoid hardcoding.